label meta_vcd_metadata_vmware.cloud.director.ui.badges  is not a valid label name

Hi, while using VCD-SD exporter, find this bug.

Discover service from **prometheus-vcd-sd**
```
 {
        "targets": [
            "10.12.146.50"
        ],
        "labels": {
            "__address__": "10.12.146.50",
            "__meta_vcd_id": "urn:vcloud:vm:02217ed9-ea98-495e-a577-d1a517017611",
            "__meta_vcd_name": "k8sw-dev-cubone",
            "__meta_vcd_network_k8s_net": "10.12.146.50",
            "__meta_vcd_num_cores_per_socket": "2",
            "__meta_vcd_num_cpus": "8",
            "__meta_vcd_org": "VDC",
            "__meta_vcd_os_type": "ubuntu64Guest",
            "__meta_vcd_project": "default",
            "__meta_vcd_status": "4",
            "__meta_vcd_storage_profile": "NonSSD",
            "__meta_vcd_vdc": "VDC"
        }
    },
    {
        "targets": [
            "10.12.147.4"
        ],
        "labels": {
            "__address__": "10.12.147.4",
            "__meta_vcd_id": "urn:vcloud:vm:12533e37-9ba1-4cd0-8919-d0360b6acdcf",
            "__meta_vcd_metadata_vmware.cloud.director.ui.badges.blue": "false",
            "__meta_vcd_metadata_vmware.cloud.director.ui.badges.green": "false",
            "__meta_vcd_metadata_vmware.cloud.director.ui.badges.magenta": "false",
            "__meta_vcd_metadata_vmware.cloud.director.ui.badges.orange": "false",
            "__meta_vcd_metadata_vmware.cloud.director.ui.badges.red": "false",
            "__meta_vcd_name": "jump-host",
            "__meta_vcd_network_srv_net": "10.12.147.4",
            "__meta_vcd_num_cores_per_socket": "1",
            "__meta_vcd_num_cpus": "2",
            "__meta_vcd_org": "VDC",
            "__meta_vcd_os_type": "ubuntu64Guest",
            "__meta_vcd_project": "default",
            "__meta_vcd_status": "4",
            "__meta_vcd_storage_profile": "NonSSD",
            "__meta_vcd_vdc": "VDC"
        }
    },
```
After this in log Prometheus contain error 

```
prometheus  | ts=2022-02-15T09:44:43.523Z caller=refresh.go:79 level=error component="discovery manager scrape" discovery=http msg="Unable to refresh target groups" err="\"__meta_vcd_metadata_vmware.cloud.director.ui.badges.green\" is not a valid label name"
``` 

docker-compose.yml
```
version: '2'

volumes:
  prometheus:

services:
  prometheus:
    image: prom/prometheus
    container_name: prometheus
    restart: always
    ports:
      - 9090:9090
    volumes:
      - prometheus:/prometheus
      - ./prometheus.yml:/etc/prometheus/prometheus.yml
      - ./service-discovery:/etc/sd

  vcd-exporter:
    image: promhippie/prometheus-vcd-sd:latest
    container_name: vcd-sd
    restart: always
    ports:
      - 9000:9000
    environment:
      - PROMETHEUS_VCD_LOG_PRETTY=true
      - PROMETHEUS_VCD_OUTPUT_ENGINE=http
      - PROMETHEUS_VCD_OUTPUT_FILE=/etc/sd/vcd.json
      - PROMETHEUS_VCD_URL=API_URL
      - PROMETHEUS_VCD_USERNAME=USER
      - PROMETHEUS_VCD_PASSWORD=PASS
      - PROMETHEUS_VCD_ORG=VDC
      - PROMETHEUS_VCD_VDC=VDC
    volumes:
      - ./service-discovery:/etc/sd
```
prometheus.yml
```
global:
  scrape_interval: 1m
  scrape_timeout: 10s
  evaluation_interval: 1m

scrape_configs:
- job_name: node
  http_sd_configs:
  - url: http://vcd-sd:9000/sd

  relabel_configs:
  - source_labels: [__meta_vcd_network_internal]
    replacement: "${1}:9100"
    target_label: __address__
  - source_labels: [__meta_vcd_name]
    target_label: instance
```

docker version: 20.10.12
docker compose version: v2.2.3
OS: Debian GNU/Linux 11 (bullseye)

I try to add key, but build crushed 

```
                                for _, entry := range metadata.MetadataEntry {
                                        normalized := strings.ToLower(
                                                strings.ReplaceAll(
                                                        entry.Key,
                                                        "-",
                                                        ".",
                                                        "_",
                                                ),
                                        )

                                        target.Labels[model.LabelName(Labels["metadataPrefix"]+normalized)] = model.LabelValue(entry.TypedValue.Value)
                                }
```
Build error:
```
pkg/action/discoverer.go:251:25: too many arguments in call to strings.ReplaceAll
        have (string, string, string, string)
        want (string, string, string)
make: *** [Makefile:99: bin/prometheus-vcd-sd] Error 2
```

Sorry, I don't have enough experience with go

--- a/changelog/unreleased/normalize-labels.md
+++ b/changelog/unreleased/normalize-labels.md
@@ -1,0 +1,7 @@
+Bugfix: Properly normalize label names
+
+We have added more character replacements for generating the label names as it
+could contain bad characters depending on the definitions within a vCloud
+Director instance. Now we are replacing `-`, `.` and `,` by `_`.
+
+https://github.com/promhippie/prometheus-vcd-sd/issues/86

--- a/hack/generate-envvars-docs.go
+++ b/hack/generate-envvars-docs.go
@@ -1,3 +1,4 @@
+//go:build ignore
 // +build ignore
 
 package main

--- a/hack/generate-labels-docs.go
+++ b/hack/generate-labels-docs.go
@@ -1,3 +1,4 @@
+//go:build ignore
 // +build ignore
 
 package main

--- a/pkg/action/discoverer.go
+++ b/pkg/action/discoverer.go
@@ -235,26 +235,12 @@ func (d *Discoverer) getTargets(ctx context.Context) ([]*targetgroup.Group, erro
 				}
 
 				for _, network := range vm.VM.NetworkConnectionSection.NetworkConnection {
-					normalied := strings.ToLower(
-						strings.ReplaceAll(
-							network.Network,
-							"-",
-							"_",
-						),
-					)
-
-					target.Labels[model.LabelName(Labels["networkPrefix"]+normalied)] = model.LabelValue(network.IPAddress)
+					normalized := normalizeLabel(network.Network)
+					target.Labels[model.LabelName(Labels["networkPrefix"]+normalized)] = model.LabelValue(network.IPAddress)
 				}
 
 				for _, entry := range metadata.MetadataEntry {
-					normalized := strings.ToLower(
-						strings.ReplaceAll(
-							entry.Key,
-							"-",
-							"_",
-						),
-					)
-
+					normalized := normalizeLabel(entry.Key)
 					target.Labels[model.LabelName(Labels["metadataPrefix"]+normalized)] = model.LabelValue(entry.TypedValue.Value)
 				}
 
@@ -292,4 +278,24 @@ func (d *Discoverer) getTargets(ctx context.Context) ([]*targetgroup.Group, erro
 
 	d.lasts = current
 	return targets, nil
+}
+
+func normalizeLabel(val string) string {
+	replaces := map[string]string{
+		"-": "_",
+		".": "_",
+		",": "_",
+	}
+
+	for original, replaced := range replaces {
+		val = strings.ReplaceAll(
+			val,
+			original,
+			replaced,
+		)
+	}
+
+	return strings.ToLower(
+		val,
+	)
 }

